### DISCUSSION TITLE
NetBoxModelCSVForm required does not allow optional parameters

### DISCUSSION THREAD

**Discussion by dainok** [netbox-community/netbox#9476] [Q&A] [answered]
### NetBox version

v3.2.3

### Python version

3.8

### Steps to Reproduce

I'm developing a NetBox plugin, with CSV import feature. I realized that using NetBoxModelCSVForm I have two options:
1. Using required=true makes the field mandatory (blank value not allowed)
2. Using required=false makes the filed forbidden (the value cannot be filled).
The CSV Import cannot be used to fill optional values (used by some objects).
Before trying to figure out how to patch it, I would like to know what is the best approach.

Mine is:
1. introduce a new attribute (optional, default is false)
2. if required and optional are true, the attribute can be filed, and a blank value is accepted
3. if required is true and optional is false, the attribute is mandatory and a blank value is not accepted
4. if required is false, optional is ignored, the attribute is ignored (it's not even mentioned in the HTML page).

Steps to reproduce:
1. Create a form using NetBoxModelCSVForm
6. create a field named optional and set required=false
7. try to import using a blank optional value

### Expected Behavior

Data should be imported with blank values.

### Observed Behavior

The post return the following error:  Row 1: Expected 1 columns but found 4

**Comment by jeremystretch** [comment]
It seems like you might just need some help writing a form, so I've moved this to a discussion. This error in particular suggests that you're not passing data in the expected format:

> Row 1: Expected 1 columns but found 4

It would help to present the code you've written thus far for someone to take a look.

**Comment by dainok** [comment]
Hello @jeremystretch I realize my report was too short. Let me explain better.

I'm building a plugin, basically I need to store access credentials (name, username, password, enable_secret). This is working fine. In the models "name" is mandatory, others are optionals (blank=True).

I would like to implement the bulk import via CSV.
I wrote a simple form using NetBoxModelCSVForm including the four fields:

~~~
class CredentialCSVForm(NetBoxModelCSVForm):
    class Meta:
        model = Credential
        fields = ('name', 'username', 'password', 'enable_password')
~~~

With this configuration if I import "test1,admin1,cisco1,cisco1" I get the error:  Row 1: Expected 1 columns but found 4 
I can import: "test1".

If i set in the form required=true on the three optional parameters I can import "test1,admin1,cisco1,cisco1" but I cannot import "test2,,,":  Row 1 username: This field is required, Row 1 password: This field is required...

~~~
class CredentialCSVForm(NetBoxModelCSVForm):
    username = forms.CharField(
        required=True
    )
    password = forms.CharField(
        required=True,
    )
    enable_password = forms.CharField(
        required=True,
    )

    class Meta:
        model = Credential
        fields = ('name', 'username', 'password', 'enable_password')
~~~

I also checked in IPAM -> VRFs and I realized that I can bulk Import VRF name but I cannot include other parameters (same error: Row 1: Expected ... columns but found...)

I tested and I think it's a bug, but maybe I'm missing something.

**Reply by candlerb** [reply]
> I also checked in IPAM -> VRFs and I realized that I can bulk Import VRF name but I cannot include other parameters (same error: Row 1: Expected ... columns but found...)

Can you give a specific example of a CSV input that fails?  The following works for me at `/ipam/vrfs/import/`, with Netbox 3.2.4:

```
name,rd,description
test123,123,this is a test
```

I note that both "rd" and "description" are optional fields.  If this doesn't work for you, it suggests your installation is broken in some other way.

**Comment by dainok** [comment]
Hello @candlerb,

Given your answer, I reinstalled a new NetBox instance using 3.2.4 release (from tar.gz), no plugins, new python venv.
Opening the bulk VRF import I can see inside the textarea:

~~~
name
~~~

If I try to import:

~~~
test123,123,this is a test
~~~

I have the following error.

~~~
 Unexpected column header "test123" found.
~~~

The same is happening with the 3.2.3 version.

I understand you can read inside the textarea:

~~~
name,rd,description
~~~

Which version are you using? Installed how?

**Reply by kkthxbye-code** [reply]
>If I try to import:
>`test123,123,this is a test`

You need to import 

```
name,rd,description
test123,123,this is a test
```

You can't just import a row, you need to define the headers.

**Comment by dainok** [comment, marked as answer]
It works thank you. I didn't realized that I could customize the header. Referring to plugins, can I customize the default expected format?
I mean:
name,username,password,enable_password

Overriding the default:
name

**Reply by candlerb** [reply]
When you upload a CSV file, the first line is the list of column headers, and the subsequent lines are the data.  You can include whichever columns you like (in whichever order you like), as long as you include the mandatory ones.

Netbox pre-fills the text area with a dummy first line which lists only the mandatory fields.  You need to replace that with a *real* set of headers that matches the *real* data in the subsequent lines.

If you are including optional fields, then it's up to you to list those optional fields in the column headings.  Otherwise, Netbox has no idea *which* optional fields you are including, or in what order.

**Reply by dainok** [reply]
Sounds good, thank you.
